### PR TITLE
Update documentation

### PR DESCRIPTION
**Pull request by AlessandroBudroni**
- Standardisation of the documentation in ./src and ./include.
- Standardisation of the header of every file in ./src and ./include
- Add prototype for the function FF_rawoutput(BIG *x,int n) in ff.c.

### CODE REVIEW

File: src/x509.c [hunk 1]
```diff
@@ -507,8 +512,6 @@ pktype X509_extract_public_key(octet *c,octet *key)
     return ret;
 }
 
-// Find pointer to main sections of cert, before extracting individual field
```
Comment (kealan) [inline src/x509.c]: Why are these removed - too verbose?

File: include/amcl.h.in [hunk 2]
```diff
@@ -2730,59 +2730,82 @@ extern void OCT_toHex(octet *src,char *dst);
...
  *
-	@param H an instance SHA256
+ *  Initialise an instance of SHA256.
+ *
+ *	@param H an instance SHA256
  */
 extern void HASH256_init(hash256 *H);
-/**	@brief Add a byte to the hash
+
+/**	@brief Process a single byte to the hash
```
Comment (kealan) [inline include/amcl.h.in]: I don't see why you changed this.

File: include/amcl.h.in [hunk 3]
```diff
@@ -2730,59 +2730,82 @@ extern void OCT_toHex(octet *src,char *dst);
...
  *
-	@param H an instance SHA384
+ *  Initialise an instance of SHA384.
+ *
+ *	@param H an instance SHA384
  */
 extern void HASH384_init(hash384 *H);
-/**	@brief Add a byte to the hash
+
+/**	@brief Process a single byte to the hash
```
Comment (kealan) [inline include/amcl.h.in]: Not sure why this was changed

File: include/amcl.h.in [hunk 4]
```diff
@@ -2730,59 +2730,82 @@ extern void OCT_toHex(octet *src,char *dst);
...
+ *  Initialise an instance of SHA512.
+ *
+ *	@param H an instance SHA512
  */
 extern void HASH512_init(hash512 *H);
-/**	@brief Add a byte to the hash
+
+/**	@brief Process a single byte to the hash
+ *
+ *  Process a single byte to the hash.
```
Comment (kealan) [inline include/amcl.h.in]: not sure why changed

File: include/amcl.h.in (hunk 2 above)
Comment (AlessandroBudroni) [inline include/amcl.h.in]: In hash.c there was "Process a single byte to the hash" while in amcl.h there was "Add a byte to the hash". So I decided to keep the same and I chose the one with Process. Same for the others

File: include/mpin.h [hunk 5]
```diff
@@ -70,322 +69,420 @@ under the License.
...
+ *  Extract pin from TOKEN for client's identity CID.
+ *
+ * 	@param sha is the hash type
+ *	@param CID is the input client identity
+ *	@param pin is an input PIN number
+ *	@param TOKEN is the client secret from which the pin is to be extracted
+ *	@return 0 or an error code
  */
-int MPIN_EXTRACT_PIN(int h,octet *ID,int pin,octet *CS);
+int MPIN_EXTRACT_PIN(int sha,octet *CID,int pin,octet *TOKEN);
```
Comment (kealan) [inline include/mpin.h]: CID -> ID  or maybe even Id as CID was meant to be Id of the client

File: include/mpin.h [hunk 6]
```diff
@@ -70,322 +69,420 @@ under the License.
...
+ *	@param TP is the input time permit
+ *	@return 0 or an error code
+ *  @note If Time Permits are disabled, set d = 0, and UT is not generated and can be set to NULL
+ *  @note If Time Permits are enabled, and PIN error detection is OFF, U is not generated and can be set to NULL
+ *  @note If Time Permits are enabled, and PIN error detection is ON, U and UT are both generated
  */
 int MPIN_CLIENT_1(int h,int d,octet *ID,csprng *R,octet *x,int pin,octet *T,octet *S,octet *U,octet *UT,octet *TP);
-/**	@brief Generate a random group element
+
+/**	@brief Generate a random group element, a random secret S
```
Comment (kealan) [inline include/mpin.h]: This is not just for the master secret so I think the original comment is better

File: include/mpin.h [hunk 7]
```diff
@@ -70,322 +69,420 @@ under the License.
...
+ *  Perform second pass of the client side of the 3-pass version of the M-Pin protocol.
+ *
+ * 	@param X an input, a locally generated random number
+ *	@param Y an input random challenge from the server
+ *	@param SEC on output = -(X+T).SEC
+ *	@return 0 or an error code
+ *  @note SEC=-(X+Y)*SEC
  */
-int MPIN_CLIENT_2(octet *x,octet *y,octet *V);
+int MPIN_CLIENT_2(octet *X,octet *Y,octet *SEC);
```
Comment (kealan) [inline include/mpin.h]: I prefer V here as this what is in the documentation. SEC is the reconstructed secret.

File: include/mpin.h [hunk 8]
```diff
@@ -70,322 +69,420 @@ under the License.
...
+ *	@param SST is the input server secret
+ *	@param W is an input random number generated by the server
+ *	@param H is an input, hash of the protocol transcript
+ *	@param HID is the hashed input client HID = H(ID)
+ *	@param xID is input from the client = x.H(ID)
+ *	@param xCID is input from the client= x.(H(ID)+H(d|H(ID)))
+ *	@param SK is the output calculated shared key
+ *	@return 0 or an error code
+ */
+int MPIN_SERVER_KEY(int sha,octet *Z,octet *SST,octet *W,octet *H,octet *HID,octet *xID,octet *xCID,octet *SK);
```
Comment (kealan) [inline include/mpin.h]: There is probably some confusion here but U = xID and UT=xCID. U and UT is what is used in documentation. So please revert to original.

File: include/mpin.h [hunk 9]
```diff
@@ -70,322 +69,420 @@ under the License.
...
- */
-int MPIN_SERVER_KEY(int h,octet *Z,octet *SS,octet *w,octet *p,octet *I,octet *U,octet *UT,octet *K);
+ *  Calculate Key on Server side for M-Pin Full.
+ *
+ *	Uses xCID internally for the key calculation, unless not available in which case xID is used
+ *	@param sha is the hash type
+ *	@param Z is the input Client-side Diffie-Hellman component
+ *	@param SST is the input server secret
+ *	@param W is an input random number generated by the server
+ *	@param H is an input, hash of the protocol transcript
```
Comment (kealan) [inline include/mpin.h]: Maybe HT instead of H

File: include/mpin.h [hunk 10]
```diff
@@ -70,322 +69,420 @@ under the License.
...
+ *	@param p is an input, hash of the protocol transcript
+ *	@param T is the input Server-side Diffie-Hellman component
+ *	@param K is the output calculated shared key
+ *	@return 0 or an error code
+ *  @note wCID = w.(A+AT)
  */
 int MPIN_CLIENT_KEY(int h,octet *g1,octet *g2,int pin,octet *r,octet *x,octet *p,octet *T,octet *K);
 
-/**	@brief AES-GCM Encryption
+/**	@brief AES-GCM Encryption of octets
```
Comment (kealan) [inline include/mpin.h]: Not sure we need to say thsi

File: src/mpin.c [hunk 11]
```diff
@@ -397,7 +403,7 @@ int MPIN_RECOMBINE_G2(octet *W1,octet *W2,octet *W)
     return res;
 }
 
-/* create random secret S */
+/* Generate a random group element, a random secret S */
```
Comment (kealan) [inline src/mpin.c]: Again not a secret

File: include/mpin.h (hunk 6 above)
Comment (AlessandroBudroni) [inline include/mpin.h]: Ok!

File: include/mpin.h (hunk 5 above)
Comment (nicolaasuni) [inline include/mpin.h]: Yes, @AlessandroBudroni in this type of PRs when a lot of comments changes there should be no changes in the code at all.

File: include/mpin.h (hunk 7 above)
Comment (AlessandroBudroni) [inline include/mpin.h]: Ok I'll restore it

File: src/mpin.c [hunk 12]
```diff
@@ -535,16 +529,13 @@ int MPIN_GET_G2_MULTIPLE(csprng *RNG,int type,octet *X,octet *G,octet *W)
     return res;
 }
 
-
-
-/* Client secret CST=s*H(CID) where CID is client ID and s is master secret */
-/* CID is hashed externally */
+/* Create a client secret in G1 from a master secret and the client ID */
```
Comment (kealan) [inline src/mpin.c]: May keep CST = s*H(Id)

File: include/mpin.h (hunk 8 above)
Comment (AlessandroBudroni) [inline include/mpin.h]: Ok!

File: src/mpin.c [hunk 13]
```diff
@@ -1041,12 +1027,13 @@ int MPIN_SERVER_KEY(int sha,octet *Z,octet *SST,octet *W,octet *H,octet *HID,oct
...
 }
 
+/* Get epoch time as unsigned integer */
 unsign32 MPIN_GET_TIME(void)
 {
     return (unsign32)time(NULL);
 }
 
-/* Generate Y = H(epoch, xCID/xID) */
+/* Generate Y=H(TimeValue,O), where TimeValue is epoch time, O is an octet, and H(.) is a hash function */
```
Comment (kealan) [inline src/mpin.c]: Y = H(epoch, U) when time permits disabled or else Y = H(epoch, UT) where U = x.H(Id) 
or UT = x.(H(Id)+H(date|H(Id)))

File: include/mpin.h (hunk 9 above)
Comment (AlessandroBudroni) [inline include/mpin.h]: ok!

Comment (AlessandroBudroni) [inline include/mpin.h]: Ok!

File: src/mpin.c (hunk 11 above)
Comment (AlessandroBudroni) [inline src/mpin.c]: Ok!

File: src/mpin.c (hunk 12 above)
Comment (AlessandroBudroni) [inline src/mpin.c]: Ok!

File: src/mpin.c (hunk 13 above)
Comment (AlessandroBudroni) [inline src/mpin.c]: /* Generate Y = H(TimeValue, xCID) when time permits disabled or else 
   Y = H(TimeValue, xCID) where xCID = x.H(Id) or xCID = x.(H(Id)+H(TimeValue|H(Id))),
   where TimeValue is epoch time, and H(.) is a hash function */

I wrote it again to make it consistent with the signature of the function
